src/Racing.cpp

```diff
@@ -97,31 +97,47 @@ static void racingEnd(CNSocket* sock, CNPacketData* data) {
     if (EPData.find(mapNum) == EPData.end() || EPData[mapNum].EPID == 0)
         return; // IZ not found
 
-    uint64_t now = getTime() / 1000;
+    EPInfo& epInfo = EPData[mapNum];
+    EPRace& epRace = EPRaces[sock];
+    // if there are no divide-by-zero dangers, and at least one factor has been specified
+    // we switch over to OG scoring
+    bool useOGScoring = (epInfo.maxPods > 0) && (epInfo.maxTime > 0) && (
```

Comment:
Not sure if there should be a separate flag in the tdata patch to use OG scoring, or if it should be inferred like this. It isn't obvious from a server admin's side when the OG scoring kicks in like this. Comments?

Reply:
I definitely don't like the idea of dynamically choosing between score formulas depending on whether one can work or not. Having it configurable in the config file is a maybe; that doesn't sound bad, but the choice should always be made explicitly.

The possibility of divide-by-zero errors in the more accurate formula just doesn't sound right. The formula (or conceptually the code around it) should compensate for that situation and do something sane in that scenario; or perhaps the TableData should be adjusted, if it's not a direct rip from the client (I don't recall ATM). Do you know which specific combinations of IZ and other factors would trip it up with the current TableData?

Reply:
There seems to be some miscommunication here on my part. To summarize:
- The values provided for the OG game's score function are discovered (and not a part of TableData etc originally), as documented here: https://github.com/FinnHornhoover/IZScoreResearch. In fact, there's nothing in the TableData that would make or break this portion of the logic.
- None of the OG game's (discovered) values would produce divide by zeros.
- My tooling for the alteration of OG game's values would not produce divide by zeros either.
- The two cases where the values I'm using here would be <= 0 are:
  - If they are not set (and thus default to 0). This would mean the values provided in the drops JSON are not meant for the OG scoring function (and we should probably disregard any settings option instructing us to use the OG scoring function).
  - If they've been malformed by the user by directly editing the text in the drops JSON. At which point, the decision mechanism is just there to weed out the edge cases that don't make sense. It would make sense to throw out an error in this case as well, it's just that the idea was "graceful failure" rather than "fast failure".

All this said, I can add a settings option for this stuff (though I would probably still need to check for the "malformed JSON" edge case). Should I go ahead with that? Or is the fields that are only involved with the OG scoring function being present a good enough setting switch? Should the server owner be responsible from a separate setting option AND the OG scoring patch, or just the OG scoring patch?

Reply:
Ah, ok. That makes way more sense.

We still don't want the racing formula changing dynamically based on what's available. If we're going to support both, it should be an explicit choice in the config file; and inappropriate or missing values in TableData should cause the server to print an error and exit when loading it in (the fail fast approach).

Also just noticed that you implemented the tdata changes as a patch file. Those were mostly meant for adapting the data to the configured client build moreso than for opting into alternate behavior. They'll be more rigid once we merge in the `patchmap` branch at some point.

Although I'm not wholly against implementing the new formula's data as an optional patch, I do think I'd prefer it to just replace the old data wholesale in the primary `drops.json` file. Then that would also imply not doing the config option and just having the new formula being the only one in further server versions. And I think that's how I'd prefer that too. We'll discuss it amongst each other some more.

Reply:
+1 on the idea of just moving to the new formula completely

Reply:
just echoing what we concluded in the discord. the next steps for this are:
- move the current scores to a "retro scores" patch file
- overwrite the current scores in drops.json with your new scores
- change the code to use your new formula by default
- - you can make it check the list of enabled patches for the "retro patch" and use the old formula if it's there. i think it would be annoying to have to enable the patch *and* set a config option, but that's just me